# Question 10: `Part (a) - (g)`

**INITIALIZATION**

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm 

carseats_data = pd.read_csv(r"C:\Users\kevin\OneDrive\Desktop\Strathmore\Classes\December 2024\Machine Learning\Assignments\Data\ALL CSV FILES - 2nd Edition\Carseats.csv")

print(carseats_data.shape)

carseats_data.head(3)


(400, 11)


Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0   9.50        138      73           11         276    120       Bad   42   
1  11.22        111      48           16         260     83      Good   65   
2  10.06        113      35           10         269     80    Medium   59   

   Education Urban   US  
0         17   Yes  Yes  
1         10   Yes  Yes  
2         12   Yes  Yes

**(a) Fit a multiple regression model to predict Sales using Price, Urban, and US**


In [2]:
# Set X and y
X = carseats_data[['Price', 'Urban', 'US']]

# Given that `urban` and `US` are categorical data, we transform that to dummy variables
X = pd.get_dummies(X, drop_first=True)
X = X.astype(int)
X = sm.add_constant(X)

y = carseats_data['Sales']

# Fit model and show summary
model = sm.OLS(y, X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.234
Method:                 Least Squares   F-statistic:                     41.52
Date:                Sat, 07 Dec 2024   Prob (F-statistic):           2.39e-23
Time:                        20:37:43   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1863.
Df Residuals:                     396   BIC:                             1879.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         13.0435      0.651     20.036      0.000      11.764      14.323
Price         -0.0545      0.005    -10.389      0.000      -0.065      -0.044
Urban_Yes     -0.0219      0.272     -0.081      0.936      -0.556       0.512
US_Yes         1.2006      0.259      4.635      0.000       0.691       1.710
==============================================================================
Omnibus:                        0.676   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.713   Jarque-Bera (JB):                0.758
Skew:                           0.093   Prob(JB):                        0.684
Kurtosis:                       2.897   Cond. No.                         628.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**(b) Provide an interpretation of each coefficient in the model. Be careful some of the variables in the model are qualitative!**

   
***Intercept:*** *`13.0435`*
- Represents the expected average sales when all predictors (Price, Urban_Yes, US_Yes) are zero. Since Price, Urban, and US cannot realistically be zero in this context, the intercept mainly serves as a baseline value for predictions.


***Price*** *`-0.0545`*
- *Interpretation:* For each unit increase in Price, the average sales decrease by `0.0545` units, holding all other variables constant.
- *Significance:* The p-value for Price is `0.000` which is very small, indicating that price is *statistically significant* in explaining variations in sales.

***Urban_Yes:*** *`-0.0219`*
- *Interpretation:* If the store is located in an urban area (Urban_Yes = 1), the average sales decrease by `0.0219` units compared to stores in non-urban areas (Urban_Yes = 0), holding all other variables constant.
- *Significance:* The p-value for Urban_Yes is `0.936`, which is much larger than the typical significance level of 0.05. This indicates that Urban_Yes is *not statistically significant* in predicting sales.

***US_Yes:*** *`1.2006`*
- *Interpretation:* If the store is located in the US (US_Yes = 1), the average sales increase by `1.2006` units compared to stores located outside the US (US_Yes = 0), holding all other variables constant.
- *Significance:* The p-value for US_Yes is `0.000`, indicating that this variable is *highly statistically significant* in explaining variations in sales.

***Overall Model Fit:***
- The **`R-squared`** value of `0.239` indicates that `23.9%` of the variability in sales is explained by the model. While this suggests that the model has limited explanatory power, the significant predictors (Price and US_Yes) provide meaningful insights.

**(c) Write out the model in equation form, being careful to handle the qualitative variables properly.**

-   *Sales = 13.0435 - 0.545(Price) - 0.0219(Urban_Yes) + 1.2006(US_Yes)*


**(d) For which of the predictors can you reject the null hypothesis H0 : #j = 0?**
- Given the p-value we reject the null hypothesis for *Price* and *Urban* as they are very small

**(e) On the basis of your response to the previous question, fit a smaller model that only uses the predictors for which there is**

In [3]:
# Using statsmodel formula for this part
import statsmodels.formula.api as smf
model1 = smf.ols(formula = 'Sales ~ Price + US', data=carseats_data)
res_model = model1.fit()
print(res_model.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     62.43
Date:                Sat, 07 Dec 2024   Prob (F-statistic):           2.66e-24
Time:                        20:37:43   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1861.
Df Residuals:                     397   BIC:                             1873.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     13.0308      0.631     20.652      0.0

**(f) How well do the models in (a) and (e) fit the data?**
- Looking at the value of R-Squared both (a) and (e) have the same value `0.239`, we can conclude that neither model fit the data very well. However, (e) is a better model compared to (a) as it uses less variables for the same level of R-Squared 

**(g) Using the model from (e), obtain 95 % confidence intervals for the coefficient(s).**


In [5]:
#Coefficients and Standard Errors
coeff_stder = {
    'Intercept': (13.0308, 0.631),
    'US_Yes': (1.1996, 0.258),
    'Price': (-0.0545, 0.005),
}

# Calculate and format confidence intervals
def calc_ci(coef, stderr, confidence=0.95):
    z_score = 2
    lower = coef - z_score * stderr
    upper = coef + z_score * stderr
    return f"[{lower:.4f}, {upper:.4f}]"

# Print output
print("95% Confidence Intervals:")
for term, (coef, stderr) in coeff_stder.items():
    interval = calc_ci(coef, stderr)
    print(f"{term}: {interval}")

95% Confidence Intervals:
Intercept: [11.7688, 14.2928]
US_Yes: [0.6836, 1.7156]
Price: [-0.0645, -0.0445]
